### PR TITLE
util/clientmetric: optimize memory layout for finding updates

### PR DESCRIPTION
**Pull request by bradfitz**
Updates #3307

### CODE REVIEW

File: util/clientmetric/clientmetric.go [hunk 1]
```diff
@@ -22,12 +22,27 @@ import (
...
-	sortedDirty bool      // whether sorted needs to be rebuilt
-	sorted      []*Metric // by name
+	numWireID   int         // how many wireIDs have been allocated
+	lastDelta   time.Time   // time of last call to EncodeLogTailMetricsDelta
+	sortedDirty bool        // whether sorted needs to be rebuilt
+	sorted      []*Metric   // by name
+	lastLogVal  []scanEntry // by Metric.regIdx
+	unsorted    []*Metric   // by Metric.regIdx
+
+	// valBackBuf is a set of free contiguous int64s to assign to Metric.v
```
Comment (josharian) [inline util/clientmetric/clientmetric.go:32]: I suggest calling this `freelist` or `int64alloc` or the like. This makes it clearer that the purpose is a custom int64 allocator.

"to assign to" isn't technically quite right--we're taking the address, not just assigning.

File: util/clientmetric/clientmetric.go [hunk 2]
```diff
@@ -22,12 +22,27 @@ import (
...
+	// Any memory address in len(valPackBuf) is free for use.
+	// They're contiguous to reduce memory pressure during diff scans.
```
Comment (josharian) [inline util/clientmetric/clientmetric.go:34]: s/memory pressure/cache churn/

**Review by josharian — APPROVED** [review]

File: util/clientmetric/clientmetric.go (hunk 1 above)
Comment (bradfitz) [reply util/clientmetric/clientmetric.go:32]: done

File: util/clientmetric/clientmetric.go (hunk 2 above)
Comment (bradfitz) [reply util/clientmetric/clientmetric.go:34]: done

**Pull request merged**